app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -91,15 +88,31 @@ def callback_options
     { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
   end
 
-  def delete_objects_of_association(main_object, association_name, worker)
-    return unless (associated_objects = main_object.public_send(association_name))
-    if associated_objects.respond_to?(:each)
-      associated_objects.each { |associated_object| delete_association_perform_later(worker, associated_object) }
+  def delete_objects_of_association(main_object, reflection)
+    deletion_worker = association_delete_worker(reflection)
+
+    if reflection.macro == :has_many
+      associated_object_ids = main_object.public_send("#{reflection.name.to_s.singularize}_ids")
+      associated_object_ids.each do |associated_object_id|
+        build_object_and_schedule_deletion(reflection, associated_object_id, deletion_worker)
+      end
     else
-      delete_association_perform_later(worker, associated_objects)
+      if main_object.respond_to?(reflection.foreign_key)
+        associated_object_id = main_object.public_send(reflection.foreign_key)
+        build_object_and_schedule_deletion(reflection, associated_object_id, deletion_worker)
+      else
+        associated_object = main_object.public_send(reflection.name)
+        delete_association_perform_later(deletion_worker, associated_object) unless associated_object
```

Comment:
As we realised yesterday in our conversation 😄 :
```suggestion
        delete_association_perform_later(deletion_worker, associated_object) if associated_object
```